File: app/src/main/java/com/example/cameraapp2/TperUtilities.java
package com.example.cameraapp2;

import android.content.Context;
import android.content.res.Resources;
import android.util.Log;

import org.apache.commons.lang3.StringUtils;
import org.osmdroid.util.GeoPoint;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TperUtilities {
    private static final String TAG = TperUtilities.class.getSimpleName();
    private final Map<Integer, String> busStopDictionary;
    private Map<Integer, GeoPoint> coordinates;
    Context context;

    public TperUtilities(Context context) {
        this.context = context;
        busStopDictionary = loadData(context);

    }

    public String getBusStopByCode(int code) {
        String stopName = busStopDictionary.get(code);
        if (stopName == null) {
            return context.getString(R.string.non_existent_bus_stop);
        }
        return stopName;
    }

    public GeoPoint getGeoPointByCode(int code) {
        return coordinates.get(code);
    }

    public List<GeoPoint> getCoupleOfCoordinatesByStopName(String stopName) {
        List<Integer> codes = getCodesByStopName(getMoreSimilarBusStop(stopName));
        List<GeoPoint> points = new ArrayList<>();

        for (int code : codes) {
            points.add(getGeoPointByCode(code));
        }
        return points;
    }

    public Boolean codeIsBusStop(String busStopCodeString) {
        try {
            int code = Integer.parseInt(busStopCodeString);
            return busStopDictionary.containsKey(code);
        } catch (NumberFormatException e) {
            Log.e(TAG, "EXCEPTION: " + e);
        }
        return false;
    }

    private Boolean isBusStop(String aPossibleBusStop) {
        return busStopDictionary.containsValue(aPossibleBusStop);
    }

    private static String findClosestMatch(Collection<String> collection, String target) {
        int currMinDistance = Integer.MAX_VALUE;
        String closest = null;

        for (String compareObject : collection) {
            int currentDistance = StringUtils.getLevenshteinDistance(compareObject, target);
            if (currentDistance < currMinDistance) {
                currMinDistance = currentDistance;
                closest = compareObject;
            }
        }

        return closest;
    }

    public String getMoreSimilarBusStop(String aPossibleBusStop) {
        Log.d(TAG, "Received:" + aPossibleBusStop);
        if (Boolean.TRUE.equals(isBusStop(aPossibleBusStop))) {
            Log.wtf(TAG, "it's a valid name for a stop");
            return aPossibleBusStop;
        } else {
            Log.wtf(TAG, "must look for similarity");
            return String.valueOf(findClosestMatch(busStopDictionary.values(), aPossibleBusStop));
        }
    }

    public Map<Integer, String> loadData(Context context) {
        Map<Integer, String> busStopList = new HashMap<>();
        coordinates = new HashMap<>();

        try {
            Resources res = context.getResources();
            InputStream in = res.openRawResource(R.raw.fermate);
            InputStreamReader isr = new InputStreamReader(in);
            BufferedReader br = new BufferedReader(isr);

            br.readLine(); //skip the first line
            String actualLine;
            while ((actualLine = br.readLine()) != null) {
                String[] data = actualLine.split(context.getString(R.string.tper_dictionary_separator));
                Integer stopCode = Integer.valueOf(data[0]);
                String stopName = data[1];
                double latitude = Double.parseDouble(data[6].replace(",", "."));
                double longitude = Double.parseDouble(data[7].replace(",", "."));
                GeoPoint stopGeoPoint = new GeoPoint(latitude, longitude);
                coordinates.put(stopCode, stopGeoPoint);
                busStopList.put(stopCode, stopName);
            }

            br.close();
            isr.close();
            in.close();

            return busStopList;

        } catch (IOException e) {
            e.printStackTrace();
            Log.d(TAG, e.toString());
        }
        return Collections.emptyMap();
    }

    public List<Integer> getCodesByStopName(String stopName) {
        List<Integer> codes = new ArrayList<>();
        for (Map.Entry<Integer, String> entry : busStopDictionary.entrySet()) {
            if (stopName.equals(entry.getValue())) {
                codes.add(entry.getKey());
            }
        }
        return codes;
    }
}

File: app/src/main/java/com/example/cameraapp2/tper/TperUtilities.kt
package com.example.cameraapp2.tper

import android.app.Application
import android.content.Context
import android.util.Log
import androidx.lifecycle.ViewModelProvider
import com.example.cameraapp2.viewmodel.BusStopViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.apache.commons.text.similarity.LevenshteinDistance
import org.osmdroid.util.GeoPoint
import kotlin.math.floor

/**
 * Utility class for TPER data.
 *
 * @constructor
 * Takes a [Context] as a parameter, used to load the TPER stops data from file.
 *
 * @param context The context of the application.
 */
class TperUtilities(context: Context) {
    private val busStopViewModel: BusStopViewModel
    private var busStopDictionary: MutableMap<Int, String> = mutableMapOf()
    private var coordinates: MutableMap<Int, GeoPoint> = mutableMapOf()

    init {
        busStopViewModel =
            ViewModelProvider.AndroidViewModelFactory.getInstance(context.applicationContext as Application)
                .create(BusStopViewModel::class.java)

        // Load data from db
        val data = busStopViewModel.readAllData

        CoroutineScope(Dispatchers.Default).launch {
            data.forEach { busStop ->
                busStopDictionary[busStop.code] = busStop.name
                coordinates[busStop.code] = GeoPoint(busStop.latitude, busStop.longitude)
            }
        }
    }

    /**
     * Returns the bus stop name from its code number
     *
     * @param code the code number of the bus stop
     * @return the name of the bus stop, or an empty string if the code is not found
     */
    fun getBusStopByCode(code: Int): String {
        return busStopDictionary[code] ?: return ""
    }

    /**
     * Returns the most similar bus stop name from a given string.
     *
     * @param aPossibleBusStop name of a possible bus stop
     * @return the name of the most similar bus stop, or an empty string if no similar bus stop is found
     */
    fun getMoreSimilarBusStop(aPossibleBusStop: String): String = when {
        aPossibleBusStop.isEmpty() -> ""
        busStopDictionary.containsValue(aPossibleBusStop) -> aPossibleBusStop
        else -> {
            Log.d(TAG, "must look for similarity")

            findClosestMatch(aPossibleBusStop) ?: ""
        }
    }

    private fun getGeoPointByCode(code: Int): GeoPoint? {
        return coordinates[code]
    }

    fun getCoupleOfCoordinatesByStopName(stopName: String): List<GeoPoint> {
        val codes = getCodesByStopName(getMoreSimilarBusStop(stopName))
        val points: MutableList<GeoPoint> = mutableListOf()
        var point: GeoPoint?
        for (code in codes) {
            point = getGeoPointByCode(code)
            if (point != null) points.add(point)
        }
        return points
    }

    fun getMostSimilarStopCode(aPossibleBusCode: String): String {
        val processedCode: String =
            NumbersInferencePostProcessing.substitutionStep(aPossibleBusCode)
        return Companion.findClosestMatch(
            busStopDictionary.keys.map { it.toString() },
            processedCode,
            LEVENSHTEIN_MAX_DISTANCE_FOR_STOP_CODE_SEARCH
        ) ?: ""
    }

    fun getCodesByStopName(stopName: String): List<Int> {
        Log.d(
            TAG, "getCodesByStopName - Received:$stopName"
        )
        val codes: MutableList<Int> = ArrayList()
        for ((key, value) in busStopDictionary) {
            if (stopName == value) {
                codes.add(key)
            }
        }
        Log.d(
            TAG, "getCodesByStopName - Found:$codes"
        )

        return codes
    }

    /**
     * Returns the most similar bus stop name to [aPossibleBusStop] using the Levenshtein distance.
     *
     * [aPossibleBusStop] is compared in uppercase, since the bus stop names are all uppercase,
     * and we don't want to consider the case in the Levenshtein distance.
     *
     * @param aPossibleBusStop
     * @return the most similar bus stop name, or null if no similar bus stop is found
     */
    private fun findClosestMatch(aPossibleBusStop: String): String? {
        return Companion.findClosestMatch(
            busStopDictionary.values,
            aPossibleBusStop.uppercase(),
            maxLevenshteinDistanceFromTarget(aPossibleBusStop)
        )
    }

    companion object {
        private val TAG = TperUtilities::class.java.simpleName

        /**
         * Levenshtein Max Distance Ratio, i.e. the max ratio between the Levenshtein distance and
         * the length of the target string.
         */
        private const val LEVENSHTEIN_MAX_DISTANCE_RATIO = 0.6

        /**
         * Maximum Levenshtein distance for a possible stop code from an actual stop code to be
         * considered a match.
         */
        private const val LEVENSHTEIN_MAX_DISTANCE_FOR_STOP_CODE_SEARCH = 1

        /**
         * Returns the maximum Levenshtein from [target] allowed for a string to be considered a
         * valid match.
         *
         * @param target The target string.
         * @return The maximum Levenshtein distance from [target].
         */
        private fun maxLevenshteinDistanceFromTarget(target: String): Int {
            return floor(target.length * LEVENSHTEIN_MAX_DISTANCE_RATIO).toInt()
        }

        /**
         * Searches for the closest match of a string to be searched in a list of strings.
         * It uses the Levenshtein distance algorithm, with a maximum distance obtained by
         * [maxLevenshteinDistanceFromTarget].
         *
         * @param collection The collection in which to search for the closest match.
         * @param toSearch The target string.
         *
         * @return The closest match, i.e the string in [collection] that is the closest to
         * [toSearch], or null if no match is found.
         */
        private fun findClosestMatch(
            collection: Collection<String>, toSearch: String, maxDistance: Int
        ): String? {
            if (toSearch.isEmpty() or collection.isEmpty()) return null

            var currMinDistance = Int.MAX_VALUE
            var closest: String? = null
            for (compareObject in collection) {
                val currentDistance =
                    LevenshteinDistance.getDefaultInstance().apply(compareObject, toSearch)
                if (currentDistance < currMinDistance) {
                    currMinDistance = currentDistance
                    closest = compareObject
                }
            }
            Log.i(TAG, "Min distance found: $currMinDistance; max allowed: $maxDistance")

            if (currMinDistance > maxDistance) closest = null
            return closest
        }
    }
}